lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.s3;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import io.trino.filesystem.FileIterator;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoOutputFile;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.RequestPayer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.partition;
+import static com.google.common.collect.Multimaps.toMultimap;
+import static java.util.Objects.requireNonNull;
+
+final class S3FileSystem
+        implements TrinoFileSystem
+{
+    private final S3Client client;
+    private final S3Context context;
+    private final RequestPayer requestPayer;
+
+    public S3FileSystem(S3Client client, S3Context context)
+    {
+        this.client = requireNonNull(client, "client is null");
+        this.context = requireNonNull(context, "context is null");
+        this.requestPayer = context.requestPayer();
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location path)
+    {
+        return new S3InputFile(client, context, new S3Location(path), null);
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location path, long length)
+    {
+        return new S3InputFile(client, context, new S3Location(path), length);
+    }
+
+    @Override
+    public TrinoOutputFile newOutputFile(Location path)
+    {
+        return new S3OutputFile(client, context, new S3Location(path));
+    }
+
+    @Override
+    public void deleteFile(Location path)
+            throws IOException
+    {
+        path.verifyValidFileLocation();
+        var uri = new S3Location(path);
+        var request = DeleteObjectRequest.builder()
+                .requestPayer(requestPayer)
+                .key(uri.key())
+                .bucket(uri.bucket())
+                .build();
+
+        try {
+            client.deleteObject(request);
+        }
+        catch (SdkException e) {
+            throw new IOException("Failed to delete file: " + path, e);
+        }
+    }
+
+    @Override
+    public void deleteDirectory(Location path)
+            throws IOException
+    {
+        var iterator = listFiles(path);
```

Comment:
I'd use the type name here `FileIterator` otherwise you find yourself asking "why didn't he use the partition method below?"